### PR TITLE
Improve non-root path handling

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

kubevirt is reducing required privileges on virt-launcher pods. The latest work in that direction is the NonRoot feature, where basically no additional privileges are required. In order to prepare virt-handler for this change, path traversal on virt-handlers privileged operations will be improved, since as soon as virt-launcher does not require additional privileges, "normal" users can enter the virt-launcher pod. Once in the pod, they can move paths around, create symlinks and other things. While this is not recommended, virt-handler has to be prepared to cope with such actions on paths which are partially owned by users.

For that purpose a `safepath` package is introduced. It adds `safepath.Path` and `safepath.File` types which it is safe to operate with. It builds an abstraction around all kinds of `unix.*At()` file functions. These abstractions will not follow symlinks and always preserve an initial root context.

In some cases it is necessary to access a `safepath.Path` object in unsafe ways. For example passing paths over commandline, where the  receiving tool will then re-create the safe path. For that purpose an `unsafepath` package is introduced. It encapsulates all ways to access such unsafe properties and can be tightly controlled by [banncheck](https://github.com/google/go-safeweb/blob/master/cmd/bancheck/doc.go).

**safepath.Path**

A successfully created `safepath.Path` is a resolved real path (no symlinks) at the time of creation. If any symlinks appear on later traversal, they are known to be unintentional or malicious. For example: `safepath.Path` can be passed to `safepath.OpenAt` to create a `safepath.File` object. This file object has an open filedescriptor. The open filedescriptor will lead to the existence of a filedescriptor-pseudo-file in `/proc/self/fd/*`. This pseudo-file is strongly connected with the underlying file. Even moving around the file will not remove the association. This SafePath can be passed to system calls which requires a path and can't take a file descriptor (like the mount command). Other `safepath.*At` functions exist too and work in a similar fashion.

The general mechanism is pretty nice explained in this post: https://blogs.gnome.org/jamesh/2018/04/19/secure-mounts/

In order to create a `safepath.Path`, either use `safepath.JoinWithRelativeRoot()` to create out of arbitrary paths a real path, or use `safepath.NewFile()` to create a path object out of a given real and already resolved path. In both chases, the returned objects are safe to use, or the initialization will fail already.

Affected places:

 * virt-chroot will get for its mount operations resolved real paths. These paths are then opened with `safepath.NewFile` to get a safe handle and passed to the actual mount syscalls.
 * containerdisks code operates fully on safe paths now
 * the isolation detector provides `safepath.Path` objects instead of simple string paths for more safe operation.
 * the hotplug code will use the safepath package too.
 * virt-api will not accept relative paths anymore, since the meaning is not clear.
 * host-disks will no longer relabel selinux labels automatically, to not unintentionally work against the intent of node admins.

**unsafepath.Path**

In order to make it in future code reviews easier to review, all attempts to get an "unsafe" string representation from a `safepath.Path` will result in calling functions from the new `unsafepath` package: `unsafepath.UnsafeAbsolute(mypath.Raw())`, `unsafepath.UnsafeRelative(mypath.Raw())` and so forth. In some places it is necessary to directly access the underlying paths, but this way it should be easier to ensure that they are only called where they should be.

**banncheck**

The [banncheck](https://github.com/google/go-safeweb/blob/master/cmd/bancheck/doc.go) linter is added to nogo to restrict the usage of the `unsafepath` package. Later it can be extended to also completely ban the usage of `path/filepath` in the node-parts of kubevirt.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Improve path handling for non-root virt-launcher workloads
```

### CODE REVIEW

**Comment by rmohr** [comment]
> thanks @rmohr, do you also think you could follow up with a brief write up that that we can put in the docs directory to point people to when people have questions about safepath usage? So after this PR is merged, follow up with basically putting bits of the PR description into a md doc so there's a reference for why we do this.

Will do that.

File: pkg/virt-handler/container-disk/mount.go [hunk 1]
```diff
@@ -361,19 +351,22 @@ func (m *mounter) Unmount(vmi *v1.VirtualMachineInstance) error {
 
 	log.DefaultLogger().Object(vmi).Infof("Found container disk mount entries")
 	for _, entry := range record.MountTargetEntries {
-		path := entry.TargetFile
-		log.DefaultLogger().Object(vmi).Infof("Looking to see if containerdisk is mounted at path %s", path)
-		if mounted, err := isolation.NodeIsolationResult().IsMounted(path); err != nil {
+		log.DefaultLogger().Object(vmi).Infof("Looking to see if containerdisk is mounted at path %s", entry.TargetFile)
+		path, err := safepath.NewFileNoFollow(entry.TargetFile)
+		if err != nil {
+			return fmt.Errorf(failedCheckMountPointFmt, path, err)
```
Comment (xpivarc) [inline pkg/virt-handler/container-disk/mount.go:357]: @rmohr FYI. Here we forgot to handle not found.

Comment (rmohr) [reply pkg/virt-handler/container-disk/mount.go:357]: :+1: thanks for the quick reaction.

**Pull request merged**